Show view api button even if there are no named api routes in the app

freddyaboulton opened the pull request:
# Description

Fixes #2160

### To test

- [x] Case of no named api routes: run `demo/calculator_blocks/run.py`

Should see the "To add api docs" page

<img width="1379" alt="image" src="https://user-images.githubusercontent.com/41651716/190218393-b75b3e94-1532-4b1b-9829-da8a656cddd5.png">

- [x] Case of named api routes: run modify `demo/calculator_blocks/run.py`   and add an api_name to `submit_btn.click` 

Should see the actual api docs

<img width="1415" alt="image" src="https://user-images.githubusercontent.com/41651716/190218564-42c96ed1-9829-457f-93c4-aa62400b1b19.png">

- [x] Case of suppressing the api page when there are named api routes: Set `show_api` to False in the above modified `calculator_blocks` demo

<img width="1404" alt="image" src="https://user-images.githubusercontent.com/41651716/190219920-b6be1608-85c7-4759-be4e-5a641baeda9f.png">

You can also do this with an interface demo like `calculator`


# Checklist:

- [x] I have performed a self-review of my own code
- [x] My code follows the style guidelines of this project
- [x] I have commented my code in hard-to-understand areas
- [x] I have made corresponding changes to the documentation
- [x] I have added tests that prove my fix is effective or that my feature works
- [x] New and existing unit tests pass locally with my changes

freddyaboulton commented on ui/packages/app/src/ApiDocs.svelte:
```diff
@@ -16,99 +16,121 @@
+			{/each}
+		</div>
+	{:else}
+		<h2 class="text-3xl text-center mb-6">
```
Big diff but I just added an if else, where the else clause is the case of no named api routes.

freddyaboulton commented on ui/packages/app/src/ApiDocs.svelte:
```diff
@@ -16,99 +16,121 @@
+			</span>
+		</h2>
+		<div>
+			To expose an api endpoint of your app in the documentation, set the <span
```
Happy to change the text/styling as well

freddyaboulton commented:
Thanks a really good point @aliabid94 ! 

I wanted to target the "empty state" of the api docs at developers since we get lots of questions from developers asking why the api docs are "missing" from their app. 

That being said, maybe it will be better if developers can disable the documentation via a parameter of the Blocks `launch` method. That way, when they see the "empty state" api docs page, they can decide to either name some of the end points or disable the documentation entirely.

abidlabs commented:
Agreed on adding a `show_api` parameter to `launch()` to allow developers to hide this page if they want. 

For the message itself, I think we should point developers to the `api_name` parameter in the Guides: https://gradio.app/sharing_your_app/#api-page

So I would suggest the message to read:

--

To expose an API endpoint of your app in these API docs, set the _api_name_ parameter of the event listener. For more information, see the ["API Page" section in the Guides](https://gradio.app/sharing_your_app/#api-page). To hide this API page, set  to _show_api=False_ in the `.launch()` method.

freddyaboulton commented on ui/packages/app/src/Blocks.svelte:
```diff
@@ -35,6 +35,7 @@
 	export let target: HTMLElement;
 	export let id: number = 0;
 	export let autoscroll: boolean = false;
+	export let show_api: boolean = true;
```
I think this is backwards compatible. I tested by manually removing `show_api` from the config and this parameter is still set to true. I think that covers us in the case you try to load a config from a demo created in a previous version that doesn't have `show_api`. 

Does this sound right to you @aliabid94 ?

aliabid94 replied:
sounds good to me

freddyaboulton commented:
@aliabid94 @abidlabs I added the `show_api` parameter to the Blocks.launch method and modified the text for the "no named routes" page as suggested.

This should be good for another look!

aliabid94 reviewed (approved):
LGTM

abidlabs commented:
Same LGTM!
